### ISSUE TITLE
ttnn.sampling produces negative token_ids or bad tokens

### ISSUE THREAD

**Issue by djordje-tt** [tenstorrent/tt-metal#31321]
### Problem description
`ttnn.sampling` op is producing either negative token_ids or bad tokens. 
Here are couple of findings up to now: 

- Without calling `ttnn.sampling(logits, seed=non_zero_value,...)` in warmup, negative token_ids are observed in first 128 decode iterations on both Llama3.3-70b Galaxy and Llama3.1-8B T3K. 
- With `ttnn.sampling(logits, seed=42,...)` bug is observed much later. For Llama3.3-70b glx after ~1000 generated tokens.
- For N300 Llama3.1-8B is producing bad outputs instead of negative token_ids when executed on Galaxy but when it's executed on T3K it's producing negative token_ids. 
- Previous assumption "Only without rand_tile_init we are seeing negative token_ids is incorrect but just delayed to later tokens.
- Running "performance and batch-32" works completely fine and can't reproduce the issue while "accuracy and batch-32" is failing. 

### Steps to repro
- get T3K or 6U machine
- checkout `divanovic/ttt-sampling-bug`
- build metal
- run `MESH_DEVICE=N300 LLAMA_DIR=/proj_sw/user_dev/Llama-3.1-8B-Instruct pytest models/tt_transformers/demo/simple_text_demo.py -k "accuracy and batch-32"`

**Comment by djordje-tt** [comment]
cc: @mtairum @sraizada-tt
 
@ntarafdar @cmaryanTT, this seems to be issue with memory clobbering given the current findings. Earlier we thought it's just `rand_tile_init()` not called but it just delays the bug and we see it in later decode iterations. Could you involve someone from the ops team to help us debug sampling op?  Related issue: https://github.com/tenstorrent/tt-metal/issues/30289

**Comment by cmaryanTT** [comment]
@djordje-tt  Do you believe the problem is in the sampling OP itself, or elsewhere? Do you know which data is getting corrupted?

@nardoTT has been looking at another memory clobbering issue and may be the best person to continue this debug. @ntarafdar FYI

**Comment by nardoTT** [comment]
@djordje-tt could you also please push the branch above when possible? Thank you!

**Comment by djordje-tt** [comment]
@cmaryanTT, still debugging that. Will sync offline with @nardoTT to discuss potential next steps and revisit my list. 
@nardoTT, pushed it!

**Comment by djordje-tt** [comment]
:x: It was hard to reproduce bug with trace off so here are multiple things I tried:
- Run on T3K on above branch
    - run mesh_device=T3K llama3.1-8b passed test
    - run mesh_device=N300 llama3.1-8b produce negative tokens but without trace it's passing
    - 1L passes, disabling program cache doesn't execute in tracing due to some reading/writing during trace execution which is prohibited.
- Run on 6U
    - run llama 3.1-8b mesh_device=N300 or T3K doesn't work due to fabric error that you cannot init fabric except for all devices.
 
✅ Latest runs on main that hit negative tokens **with trace off**: 
- Run Llama3.3-70B galaxy with max_tokens_generated=1280
- Independent of seed=0 or seed=42 called first negative tokens are bug is reproduced. 
- Run model with 1L produces negative token in 5th iteration

Next steps: 
- adding DPRINTS to figure out where negative token_ids come from and pursue bug further
- Do we want to fix ttnn.sampling op for non-traced version? Per @sraizada-tt, the op is producing bad outputs for non-traced version which means we have obviously bugs in the op which might be related or completely orthogonal bug to this one.

**Comment by ekuslysTT** [comment]
Notes:
- ttnn.sampling still owned by Models
- blocking
- Once non-traced version is fixed, traced version should work, too.

Action Items @djordje-tt :
- open new issue to get non-traced version working. Then and re-test, align with @nardoTT

**Comment by djordje-tt** [comment]
I am still debugging the issue. No useful findings yet!

**Comment by ekuslysTT** [comment]
@djordje-tt PR to go for review and merge queue today/tomorrow

**Comment by djordje-tt** [comment]
Bug is fixed and  in discussion with Stuti, non-traced version of ttnn.sampling is working just fine. What does not work correctly is Llama3.3-70b in non-trace version (producing repeating tokens), @sraizada-tt  to open issue with additional context if any.

**Comment by djordje-tt** [comment]
Here is the link to related issue where PRNG state will be initialized at the device opening: https://github.com/tenstorrent/tt-metal/issues/31901
